src/Dom/Document.php

```diff
@@ -0,0 +1,968 @@
+		}
+
+		return $encoding;
+	}
+
+	/**
+	 * Find a given tag with a given attribute.
+	 *
+	 * If multiple tags match, this method will only return the first one.
+	 *
+	 * @param string $content   Content in which to find the tag.
+	 * @param string $element   Element of the tag.
+	 * @param string $attribute Attribute that the tag contains.
+	 * @return string[] The requested tags. Returns an empty array if none found.
+	 */
+	private function find_tags( $content, $element, $attribute = null ) {
+		$matches = [];
+		$pattern = empty( $attribute )
+			? sprintf(
+				self::HTML_FIND_TAG_WITHOUT_ATTRIBUTE_PATTERN,
+				preg_quote( $element, '/' )
+			)
+			: sprintf(
+				self::HTML_FIND_TAG_WITH_ATTRIBUTE_PATTERN,
+				preg_quote( $element, '/' ),
+				preg_quote( $attribute, '/' )
+			);
+
+		if ( preg_match( $pattern, $content, $matches ) ) {
+			return $matches;
+		}
+
+		return [];
+	}
+
+	/**
+	 * Find a given tag with a given attribute.
+	 *
+	 * If multiple tags match, this method will only return the first one.
+	 *
+	 * @param string $content   Content in which to find the tag.
+	 * @param string $element   Element of the tag.
+	 * @param string $attribute Attribute that the tag contains.
+	 * @return string|false The requested tag, or false if not found.
+	 */
+	private function find_tag( $content, $element, $attribute = null ) {
+		$matches = $this->find_tags( $content, $element, $attribute );
+
+		if ( empty( $matches ) ) {
+			return false;
+		}
+
+		return $matches[0];
+	}
+
+	/**
+	 * Extract an attribute value from an HTML tag.
+	 *
+	 * @param string $tag       Tag from which to extract the attribute.
+	 * @param string $attribute Attribute of which to extract the value.
+	 * @return string|false Extracted attribute value, false if not found.
+	 */
+	private function extract_value( $tag, $attribute ) {
+		$matches = [];
+		$pattern = sprintf(
+			self::HTML_EXTRACT_ATTRIBUTE_VALUE_PATTERN,
+			preg_quote( $attribute, '/' )
+		);
+
+		if ( preg_match( $pattern, $tag, $matches ) ) {
+			return empty( $matches['full'] ) ? $matches['partial'] : $matches['full'];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Sanitize the encoding that was detected.
+	 *
+	 * If sanitization fails, it will return 'auto', letting the conversion
+	 * logic try to figure it out itself.
+	 *
+	 * @param string $encoding Encoding to sanitize.
+	 * @return string Sanitized encoding. Falls back to 'auto' on failure.
+	 */
+	private function sanitize_encoding( $encoding ) {
+		if ( ! function_exists( 'mb_list_encodings' ) ) {
+			return $encoding;
+		}
+
+		static $known_encodings = null;
+
+		if ( null === $known_encodings ) {
+			$known_encodings = array_map( 'strtolower', mb_list_encodings() );
+		}
+
+		if ( array_key_exists( strtolower( $encoding ), self::$encoding_map ) ) {
+			$encoding = self::$encoding_map[ strtolower( $encoding ) ];
+		}
+
+		if ( ! in_array( strtolower( $encoding ), $known_encodings, true ) ) {
+			return self::UNKNOWN_ENCODING;
+		}
```

Comment:
Suggested:

```php
		$encoding = strtolower( $encoding );
		if ( isset( self::$encoding_map[ $encoding ] ) ) {
			$encoding = self::$encoding_map[ self::$encoding_map[ $encoding ] ];
		}

		if ( ! in_array( $encoding, $known_encodings, true ) ) {
			return self::UNKNOWN_ENCODING;
		}
```

Reduce calls to `strtolower()` and `isset()` is marginally faster than `array_key_exists()`, last I checked.

Reply:
Yes, that's better (apart from the double look-up). `isset()` is not technically the same, but still good to use here.

Reply:
Oh, but `$encoding` itself needs to stay in its original form. I'll adapt.

Reply:
Resolved via https://github.com/ampproject/amp-wp/pull/3758/commits/e94e3caa06d321a56b0e394ba7e7effc4e2471e8